### ISSUE TITLE
Custom coercions

### ISSUE THREAD

**Issue by jodosha** [lotus/model#36]
**Status:** In Progress
**Taker:** @midas

Allow the mapper to accept custom coercers.

Given the following SQL table
```sql
CREATE TABLE products (
  id      serial primary key,
  price   integer NOT NULL,
);
```
```ruby
Money = Struct.new(:amount) do
  def to_int
    amount
  end
end

class Product
  include Lotus::Entity
  self.attributes = :price
end

Lotus::Model::Mapper do
  collection :products do
    entity Product
    attribute :id, Integer
    attribute :price, Money # this coerces the integer value to a Money instance
  end
end

class ProductRepository
  include Lotus::Repository
end

product = Product.new(money: Money.new(100))
ProductRepository.persist(product) # => price must be persisted as an integer (see the schema) 

product = ProductRepository.find(product.id)
product.price # => #<struct Money amount=100>
```

**Comment by midas** [comment]
@jodosha In the example you define `#to_int` on the `Money` struct.  Can you explain to me why you chose `#to_int` instead of `#to_i`?  What method would need to be defined if the coercion was to a string?  

Are you trying to avoid naming collisions with the accepted convention of `#to_i`, `#to_f`, `#to_s`, `#to_a` and `#to_h`?  If so, should we be more explicit on the intention of the method?  Something like `#as_persisted_int`?

**Comment by midas** [comment]
@jodosha @joneslee85 I have an initial pass at my implementation working for the custom coercions, however, I am running into a wall on the primitive coercions.  Because there is not a 1-to-1 mapping between Ruby primitives and DB primitives, I am having trouble determining if an attribute is primitive or custom from the adapter module.  See https://github.com/midas/model/commit/16a0fb97dbc7f395e1c7cae682e82e7d4c50b7b2#diff-5a44340703af0a85449e7cc15bfc2076R220.

From within the mapping module it is easy to determine the primitive coercions: https://github.com/midas/model/commit/16a0fb97dbc7f395e1c7cae682e82e7d4c50b7b2#diff-54737b3eb381390bbd74760bb86b1932R54.

Thus, I need to find a way to determine if the coercion is primitive or not in the adapter.  Not wanting to couple things that should not be, I am looking for guidance at this point as my understanding of the architecture of this library is still rather limited.

Several ideas:

* Move Coercer#primitive_coercion? to Coercions and access it directly from the adapter.  This seems bad to me as it couples the adapter and mapper?)
* Include this information somewhere in the @mapped_collection that is already available to the Lotus::Model::Adapters::Sql::Collection.  This seems much better, but the question becomes where?  Do we include a list of primitive coercions available?  Do we include a list of known custom coercions in the attributes?

Any guidance you can provide is appreciated.  I am almost there with this feature (excluding any refactoring that flows from a code review).  I think this is the last major problem to solve.  Thanks!

**Comment by jodosha** [comment]
@midas Thank you for taking care of this. Have you tried if the code from `Lotus::Validations` does the trick? https://github.com/lotus/validations#custom-coercions

**Comment by midas** [comment]
@jodosha I have not tried the code from Lotus::Validations yet.  I will give it a look.  However, I am a little confused.  Mentioned prior in this thread:

>The mapper isn't aware of the database schema types for each column, but this information can be easily retrieved. Anyway I'd push this responsibility to the adapter, which is closer to the database.

>The example of this issue depicts a scenario where an SQL database is used, so we are using concepts like integers, taking for grant that the database will understand them.

>Now, imagine for a second a database that only understand strings, integer would be an advanced data type for it. In this case, the adapter should provide this feature, in order to let the mapper to safely use Integer

It does not seem to me that the code in Lotus::Validations could possibly honor the guidance in the quote above as it probably does not interface with the mapper.  Am I getting the issue of coercion confused with the more advanced feature spoken about in the thread?  Any guidance to clarify is greatly appreciated.

Thanks!

**Comment by jodosha** [comment]
### Initial idea

I've initially suggested that feature of custom coercions when they were used only in one direction from the database to Ruby land. It was easy then an similar to Lotus::Validations: take a scalar value (`23`) and instantiate a custom Ruby type (`FavNumber`).

### Current limitations

Then for data integrity reasons we needed to do a two ways coercion too. Please have a look at #138 and #141. 

The problem with this change is that a custom Ruby type cannot be easily understood from the database adapters like `pg` or `sqlite3` gems. Eg. `FavNumber.new('23')` What to do with this object?

### Proposal

Imagine the follow custom Ruby type and a this mapping:

```ruby
class FavNumber
  def initialize(number)
    @number = number.to_i
  end

  def to_int
    @number
  end
end
```

```ruby
collection :users do
  entity User
  attribute :fav, FavNumber
end
```

Imagine now that in the database the column `users.fav` is an integer.
When we coerce an entity to trasform it into a serialized value, instead of use the mapped type (`FavNumber`) we should use the database type (`Integer`).

Please have a look at https://github.com/lotus/model/blob/master/lib/lotus/model/mapping/coercer.rb#L61 and https://github.com/lotus/model/blob/master/lib/lotus/model/mapping/coercer.rb#L63.

Let me translate in Ruby terms:

```ruby
fav = FavNumber.new('23')
user = User.new(fav: fav)

# Corcer should internally have a look at the database type for `fav` and use the correct `Lotus::Utils::Kernel` method to coerce it instead of the type defined in the mapper.

Coercer.new.to_record(user) # => { fav: 23 }
```

### Technical notes

* An adapter should be able to understand database supported types and to find a proper `Utils::Kernel` primitive coercer. Eg. for a timestamp use `Kernel.DateTime`.

* The custom type (`FavNumber`) MUST expose the right explicit Ruby type conversion protocol: eg. `#to_int`, `#to_str`, `#to_h`, which are understood by `Utils::Kernel`.

* `Lotus::Model::Mapping::Coercer` could have the need of be aware of the database schema and this would increase its internal complexity.

* Non typed databases MUST serialize data with `#to_str`, think of `Memory` or `FileSystem` adapters.

### Conclusion

If this feature requires to increase complexity or add too much runtime overhead, I'm keen to drop it.

**Comment by jodosha** [comment]
@midas Please ping me in chat if the explanation above isn't clear or if you have other questions.

**Comment by pascalbetz** [comment]
@jodosha @AlfonsoUceda 

I created a proof of concept on how this could be done. Feedback welcome (i went a bit a different road). If you are interested in investigating this approach a bit further, then send me a message.

Commit + some comments are here: https://github.com/pascalbetz/lotus-model/commit/a033e6c15e37a8c6b4f89a66f97879f234e52da1

Could solve #183 and #136 as well, see examples + tests.

**Comment by pascalbetz** [comment]
Forgotten to explain the idea a bit:

I added _Type_ classes that represent either a _Atom_ or a _Composite_ attribute of an entity. They can form a hierarchy (for nested/composite attributes). And they can map an entity from/to a record. 
This moved a lot of code out of Coercer and into those Types.

One can "easily" write a custom type (lets say your class does not provide writers). 

Also one can use these types to define a _default value_ (currently predefined when mapping is setup, but one could imagine a lambda doing "dynamic default values" (Time.now). See #136.

By providing a custom type one can also define a custom coercion mechanism (see #183) 

I have not looked into it from a performance point of view (might be harder to precompile such a structure). 

So far the changes are non-breaking (or could be non-breaking...currently some tests are failing or just have been disabled).

Have i mentioned that i like it? :grinning:

**Comment by pascalbetz** [comment]
Also looked into the failing tests regarding dirty tracking / repository.

DirtyTracking expects the initial state to be set using the _initializer_. Currently my approach sets the values from the record using the _setters_.

**Comment by pascalbetz** [comment]
I created a PR on my clone so one can easily see the diff and i have place to add all the description...

https://github.com/pascalbetz/lotus-model/pull/1

**Comment by jodosha** [comment]
@pascalbetz Thanks for the PR and the explanation. I'm going to leave my comments over there.

**Comment by pascalbetz** [comment]
@jodosha thanks for the feedback. As mentioned this is not a PR meant to be integrated but more to discuss the idea.

**Comment by jodosha** [comment]
Implemented by #216
